### DISCUSSION TITLE
Uploading DX files

### DISCUSSION THREAD

**Discussion by schmalgo** [molstar/molstar#153] [Q&A] [answered]
Thank you for developing this powerful visualization tool!
I hope you can help us with this: Similar to #68 we would like to upload multiple DX files and present them as IsoSurfaces (originially we tried the pdbe-molstar plugin - but it apparently can't load dx files) in different colors. I tested the following code with the viewer example. This code seems to be working - at least to upload the data:

`dxurl='myfile.dx.gz'`
`viewer.loadStructureFromUrl(dxurl, 'dx', true)`
 * The data will be uploaded, but how can I (...without using the viewer buttons)
   * present it as IsoSurface
   * choose the color
   * change the label
 * Since we have volume data here, I can't use the a solution like the one in #68 , right?
 * Along this line - this routine doesn't return the data - wouldn't it be neat so that one can manipulate it?

**Comment by dsehnal** [comment, marked as answer]
There is this function https://github.com/molstar/molstar/blob/master/src/apps/viewer/index.ts#L244

You can do all the things you mention by reusing/modifying this code. Let me know if you need more help.

**Reply by schmalgo** [reply]
Wow, thanks for the hint! So I tried this and it works:
```
dxurl='myfile.dx.gz'
viewer.loadVolumeFromUrl(dxurl, 'dx', false, [ {
    type: 'relative',
    value: 2,
    color: 0x00ff00
    alpha: 1,
}], 'my dx' )
```

One minor additional question: How can i later highlight / focus to this volume (e.g. create a button event to highlight it?)

**Reply by dsehnal** [reply]
Something like this 

```ts
const cell = this.plugin.state.data.selectQ(q => q.ofType(PluginStateObject.Volume.Representation3D))[0];
if (cell) {
  PluginCommands.Interactivity.Object.Highlight(this.plugin, { state: this.plugin.state.data, ref: cell.transform.ref });
}
// ...
PluginCommands.Interactivity.ClearHighlights(this.plugin);
```

should work.

Depending on your use case, you might have to create the state yourself so that you can store selectors or assign refs to your individual volumes to better manipulate them.

But if you just have multiple volumes, using ``[index]`` instead of ``[0]`` in the cell selector should do the trick.

**Reply by schmalgo** [reply]
Uh, frankly saying I am lost here - can I access e.g. PluginStateObject or PluginCommands from Javascript? 
Your first code snippet was easy to convert. But here?

So I started with something like this (which obviously doesn't work):
```const cell = viewer.plugin.state.data.selectQ(q => q.ofType(...```

**Reply by dsehnal** [reply]
No, unfortunately you need to bring the full might of "modern" web development with webpack and that jazz for it to work.

We might look into providing plain-JS-accessible version of Mol* too in the future -- maybe it isn't hard to do.
